File: Tests/ElevateCoreTests/ClaimsChallengeTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct ClaimsChallengeTests {
    @Test func extractsAndDecodesClaims() {
        let json = #"{"access_token":{"acrs":{"essential":true,"values":["c1"]}}}"#
        let b64 = Data(json.utf8).base64EncodedString().trimmingCharacters(in: CharacterSet(charactersIn: "="))
        let header = #"Bearer realm="", authorization_uri="https://login.microsoftonline.com/common/oauth2/authorize", error="insufficient_claims", claims="\#(b64)""#
        #expect(ClaimsChallenge.parse(wwwAuthenticate: header) == json)
    }

    @Test func returnsNilWithoutClaims() {
        #expect(ClaimsChallenge.parse(wwwAuthenticate: #"Bearer realm="""#) == nil)
    }

    @Test func interactionRetryRunsInteractiveThenRetries() async throws {
        let tokens = FakeTokenProvider()
        let identity = Identity(id: "i", upn: "u@x", displayName: "U", homeTenantId: "t")
        var calls = 0
        let result = try await InteractionRetry.run(tokens: tokens, identity: identity, tenantId: "t", scopes: GraphScopes.all) {
            calls += 1
            if calls == 1 { throw PIMError.claimsChallenge("{}") }
            return "ok"
        }
        #expect(result == "ok")
        #expect(calls == 2)
        #expect(await tokens.interactiveCalls.count == 1)
        #expect(await tokens.interactiveCalls.first?.claims == "{}")
    }
}

File: windows/tests/Elevate.Core.Tests/ClaimsChallengeTests.cs
using System.Text;
using Elevate.Core.Networking;
using FluentAssertions;

namespace Elevate.Core.Tests;

public class ClaimsChallengeTests
{
    [Fact]
    public void Parse_ExtractsAndDecodesClaims()
    {
        const string json = """{"access_token":{"acrs":{"essential":true,"values":["c1"]}}}""";
        var b64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(json)).TrimEnd('=');
        var header =
            $"""Bearer realm="", authorization_uri="https://login.microsoftonline.com/common/oauth2/authorize", error="insufficient_claims", claims="{b64}" """;

        ClaimsChallenge.Parse(header).Should().Be(json);
    }

    [Fact]
    public void Parse_ReturnsNullWithoutClaims()
    {
        ClaimsChallenge.Parse("""Bearer realm="" """).Should().BeNull();
    }

    [Fact]
    public void Parse_ReturnsNullForUndecodableClaims()
    {
        ClaimsChallenge.Parse("""Bearer claims="!!!not base64!!!" """).Should().BeNull();
    }

    [Fact]
    public void MultiFactor_RequestsAnMfaAmrValue()
    {
        ClaimsChallenge.MultiFactor.Should().Be("""{"access_token":{"amr":{"values":["mfa"]}}}""");
    }

    [Fact]
    public void AuthenticationContext_EmbedsTheEscapedContextId()
    {
        ClaimsChallenge.AuthenticationContext("c1")
            .Should().Be("""{"access_token":{"acrs":{"essential":true,"value":"c1"}}}""");

        ClaimsChallenge.AuthenticationContext("""a"\b""")
            .Should().Be("""{"access_token":{"acrs":{"essential":true,"value":"a\"\\b"}}}""");
    }
}
